timelim.1
.\" Copyright © 2020-2021 Johnothan King. All rights reserved.
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in all
.\" copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
.\" SOFTWARE.
.\"
.Dd December 7, 2021
.Dt TIMELIM 1
.Os
.Sh NAME
.Nm timelim
.Nd A program capable of setting ridiculous time limits
.Sh SYNOPSIS
.Nm
.Op Fl jsSvV
number[suffix]
.Sh DESCRIPTION
.Nm Timelim
is an implementation of the
.Nm sleep(1)
utility capable of setting time limits as short as a
few nanoseconds or as long as multiple centuries.
.Nm Timelim
will accept floating point arguments, but they will be
processed as integers internally with the decimal being used as a delimiter.
.sp
.Nm Timelim
accepts the following flags:
.sp
.Nm -j, --julian
This causes
.Nm Timelim
to use the Julian calendar instead of the Gregorian calendar.
.sp
.Nm -s, --signal
.Nm Timelim
will function as normal until it receives a POSIX signal or SIGINFO,
which will then cause it finish execution with an exit status of zero.
If no argument is given or the argument is zero,
.Nm Timelim
will wait indefinitely wait for a signal without timing out.
This option was implemented for compatibility with the
.Nm ksh93(1)
sleep builtin.
.sp
.Nm -S, --sidereal
This causes
.Nm Timelim
to use the Sidereal year instead of the Gregorian year.
.sp
.Nm -v, --verbose
This will cause
.Nm Timelim
to print out the number of seconds and nanoseconds it will wait for,
then output `Time's up!` when finished.
.sp
.Nm -V, --version
Prints the version of
.Nm Timelim
currently installed on the machine.
.sp
.Nm -?, --help
Shows usage information regarding
.Nm Timelim .
.sp
.Nm Timelim
will also accept the following suffixes appended to arguments:
.sp
.Nm n
Nanoseconds (cannot be a floating point)
.sp
.Nm u
Microseconds (cannot be a floating point)
.sp
.Nm l
Milliseconds (cannot be a floating point)
.sp
.Nm s
Seconds (same as no suffix)
.sp
.Nm m
Minutes
.sp
.Nm h
Hours
.sp
.Nm d
Days
.sp
.Nm w
Weeks
.sp
.Nm f
Fortnights (2 weeks)
.sp
.Nm o
Months (equal to `year / 12`, see below)
.sp
.Nm y
Denotes years (defaults to 31556952 seconds, passing
.Nm --sidereal
changes this to 31558150 seconds, passing
.Nm --julian
changes this to 31557600 seconds)
.sp
.Nm x
Decades
.sp
.Nm c
Centuries
.sp
.Nm a
Millennia
.sp
If
.Nm Timelim
is sent SIGPWR or SIGINFO, it will show the user how
many seconds and nanoseconds are remaining until
.Nm Timelim
finishes execution.
.Sh EXAMPLES
Sleep for 2 seconds with verbose output (when run as sleep):
sleep -v 2 or sleep -v 2s
.sp
Sleep for 4 months, 1 fortnight, 2 weeks and 2000
nanoseconds (with verbose output, suffixes):
timelim --verbose 4m 1f 2w 2000n
.sp
Sleep for 7 millennia and a month, using the Julian calendar:
timelim -j 7a 1o
.sp
Set a time limit to last a minute unless a signal is received:
timelim -s 1m or ./timelim -s 60s
.sp
Set a time limit to last one sidereal year and three weeks:
timelim -S 1y 3w
.sp
Set a time limit to only last 24000 nanoseconds:
timelim 24000n
.sp
Set a time limit to last 3.6 centuries, 2.4 decades and 4.1 months:
timelim 3.6c 2.4x 4.1o
.Sh SEE ALSO
nanosleep(2), sleep(3), signal(7), ksh93(1)
.Sh AUTHOR
Johnothan King


timelim.c
/*
 * Copyright © 2018-2020 Johnothan King. All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/*
 * timelim -- A program capable of setting very long time limits
 * This program can function as a replacement for sleep(1)
 */

#if defined(__TINYC__)
#undef _FORTIFY_SOURCE // Silence warnings
#endif

#include <assert.h>
#include <errno.h>
#include <getopt.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdnoreturn.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

// Timelim's version number
#define TIMELIM_VERSION "v3.1.0"

// Macros for compiler optimization
#define cold        __attribute__((__cold__))
#if !defined(__TINYC__)
#define likely(x)   (__builtin_expect((x), 1))
#define unlikely(x) (__builtin_expect((x), 0))
#else
#define likely(x)                (x)
#define unlikely(x)              (x)
#define __builtin_unreachable()
#endif

/*
 * Define the number of nanoseconds wasted during execution to subtract from the total time to sleep
 * This number is rather conservative, most machines will benefit from increasing the OVERHEAD_MASK
 */
#ifndef OVERHEAD_MASK
#define OVERHEAD_MASK 330000
#endif

// Colors
#define CYAN  "\x1b[1;36m"
#define WHITE "\x1b[1;37m"
#define RESET "\x1b[m"

// Macros defining the length of various units of time (weeks, years, etc.) in seconds.
#define MINUTE    60
#define HOUR      3600
#define DAY       86400
#define WEEK      604800
#define FORTNIGHT 1209600

// Universal variables
static int current_signal = 0;
extern char *__progname;

// Display usage of Timelim
static cold noreturn void usage(void)
{
    // Usage info
    printf("Usage: %s [-jsvV?] number[suffix] ...\n"
           "  -j, --julian     Use the Julian calendar instead of the Gregorian calendar\n"
           "  -s, --signal     Sleep until Timelim receives a signal or times out\n"
           "  -S, --sidereal   Use the Sidereal year instead of the Gregorian year\n"
           "  -v, --verbose    Enable verbose output\n"
           "  -V, --version    Show Timelim's version number\n"
           "  -?, --help       Display this text\n",
           __progname);
    exit(1);
}

// Print the number of seconds and nanoseconds remaining
static void nprint(unsigned long length, const char *unit)
{
    if unlikely (length == 1) // While `sleep 1` is common, `sleep -v 1` is not
        printf("%lu %s", length, unit);
    else
        printf("%lu %ss", length, unit);
}

// This function parses all numbers after the decimal (such as 1.12 or 4.5w)
static long parse_float(const char *arg, bool suffix)
{
    // Get the radix point
    char *radix = ".";
    if (strchr(arg, ','))
        radix = ",";
    else if (!strchr(arg, '.'))
        return 0; // Not a floating point

    // Set a char variable called 'base' to the relevant position
    char *modarg, *tofree, *base;
    tofree = modarg = strdup(arg); // Avoid modification of the original argument
    assert(modarg != NULL);
    strsep(&modarg, radix);
    base = strsep(&modarg, radix);
    size_t sz = strlen(base);
    if (suffix)
        --sz;

    // Set the multiplier depending on the length of base
    long num = atol(base);
    free(tofree); // Now the duplicated string can be freed from memory
    switch (sz) {
        case 1:
            return num * 100000000;
        case 2:
            return num * 10000000;
        case 3:
            return num * 1000000;
        case 4:
            return num * 100000;
        case 5:
            return num * 10000;
        case 6:
            return num * 1000;
        case 7:
            return num * 100;
        case 8:
            return num * 10;
        case 9:
            return num;
        default:
            while (num > 999999999)
                num = num / 10;
            return num;
    }
}

// Set current_signal to the signal that was sent to Timelim
static void sighandle(int sig) { current_signal = sig; }

// Main function
int main(int argc, char *argv[])
{
    // Arguments are required
    if unlikely (argc < 2) {
        usage();
        __builtin_unreachable();
    }

    // General variables
    struct timespec timer = { 0 };
    unsigned long centuries = 0;
    bool signal_wait = false, verbose = false;
    int year = 31556952; // Gregorian year default
    const char *number;

    // Long options for getopt_long
    struct option long_opts[] = { { "julian", no_argument, NULL, 'j' },
                                  { "signal", no_argument, NULL, 's' },
                                  { "sidereal", no_argument, NULL, 'S' },
                                  { "verbose", no_argument, NULL, 'v' },
                                  { "version", no_argument, NULL, 'V' },
                                  { "help", no_argument, NULL, '?' },
                                  { NULL, 0, NULL, 0 } };

    // Parse the options
    int args;
    while ((args = getopt_long(argc, argv, "jsSvV?", long_opts, NULL)) != -1)
        switch (args) {

            // Version info
            case 'V':
                printf(WHITE "Timelim " CYAN TIMELIM_VERSION RESET "\n");
                return 0;

            // Use the Julian calendar
            case 'j':
                year = 31557600;
                break;

            // Implement ksh's sleep -s flag
            case 's':
                signal_wait = true;
                break;

            // Use the Sidereal year
            case 'S':
                year = 31558150;
                break;

            // Verbose output
            case 'v':
                verbose = true;
                break;

            // Usage (with ksh93u+ compatibility)
            case '?':
                usage();
                __builtin_unreachable();
        }

    // Parse suffixes
    int multiplier, suffix_location; // Both must be int, suffixes due to `- 1`
    bool suffix;
    argv += 1;
    while ((number = *argv++)) {
        multiplier = 1;
        suffix = true;

        // Reject zero length (-1) arguments
        suffix_location = strlen(number) - 1;
        if unlikely (suffix_location < 0) {
            usage();
            __builtin_unreachable();
        }

        // If the argument has a dash, skip it
        if (strchr(number, '-') != NULL)
            continue;

        // GNU suffix parsing with partial compatibility for ksh93u+ behavior
        switch (number[suffix_location]) {
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9':
                suffix = false;
                // FALLTHRU
            case 'S':
            case 's':
                break; // Do nothing
            case 'M':
            case 'm':
                multiplier = MINUTE;
                break;
            case 'H':
            case 'h':
                multiplier = HOUR;
                break;
            case 'D':
            case 'd':
                multiplier = DAY;
                break;
            case 'W':
            case 'w':
                multiplier = WEEK;
                break;
            case 'F':
            case 'f':
                multiplier = FORTNIGHT;
                break;
            case 'O':
            case 'o':
                multiplier = year / 12; // Different from ISO 8601
                break;
            case 'Y':
            case 'y':
                multiplier = year;
                break;
            case 'X':
            case 'x':
                multiplier = year * 10;
                break;
            case 'L':
            case 'l':
                timer.tv_nsec += atol(number) * 1000000;
                continue;
            case 'U':
            case 'u':
                timer.tv_nsec += atol(number) * 1000;
                continue;
            case 'N':
            case 'n':
                timer.tv_nsec += atol(number);
                continue;
            case 'C':
            case 'c':
                centuries += strtoul(number, NULL, 10);
                multiplier = year * 100;
                goto nano;
            case 'A':
            case 'a':
                centuries += strtoul(number, NULL, 10) * 10;
                multiplier = year * 1000;
                goto nano;
            default: // Reject invalid arguments
                printf("`%c` is not a valid suffix!\n", number[suffix_location]);
                return 1;
        }

        // Set the number of seconds and nanoseconds
        timer.tv_sec += (time_t)atol(number) * multiplier;
    nano:
        timer.tv_nsec += parse_float(number, suffix) * multiplier;
    }

    // To improve accuracy, subtract 330,000 nanoseconds to account for overhead
    if (timer.tv_nsec > OVERHEAD_MASK) {
        if likely (timer.tv_sec > 0) {
            timer.tv_sec -= 1;
            timer.tv_nsec += 1000000000 - OVERHEAD_MASK;
        } else
            timer.tv_nsec -= OVERHEAD_MASK;

        // The overhead of just executing causes inaccuracy at this point, so just set this to zero
    } else
        timer.tv_nsec = 0;

    // The number of nanoseconds cannot exceed one billion
    while (timer.tv_nsec > 999999999) {
        long esec = timer.tv_nsec / 1000000000;
        timer.tv_sec += esec;
        timer.tv_nsec -= esec * 1000000000;
    }

    // Set up the signal handler now
    struct sigaction actor;
    actor.sa_handler = sighandle;
    actor.sa_flags = 0;

    // When -s was passed, handle all POSIX signals that do not kill Timelim
    if (signal_wait) {
        sigaction(SIGCHLD, &actor, NULL);
        sigaction(SIGCONT, &actor, NULL);
        sigaction(SIGQUIT, &actor, NULL);
        sigaction(SIGTSTP, &actor, NULL);
        sigaction(SIGURG, &actor, NULL);
    }

    // Handle SIGINFO or SIGPWR, depending on which is available
#ifdef SIGINFO
    sigaction(SIGINFO, &actor, NULL);
#else
#define SIGINFO 0
#endif
#ifdef SIGPWR
    sigaction(SIGPWR, &actor, NULL);
#else
#define SIGPWR 0
#endif

    // Wait indefinitely if -s was passed without a defined timeout
    if (signal_wait && timer.tv_sec == 0 && timer.tv_nsec == 0) {
        if (verbose)
            printf("Waiting for a signal...\n");
        pause();
        return 0; // pause(2) does not return 0, so it must be done separately
    }

    // Print out the number of seconds to sleep
    if (verbose) {
        printf("Sleeping for ");
        nprint(centuries * ((unsigned long)year * 100) + (unsigned long)timer.tv_sec, "second");
        printf(" and ");
        nprint((unsigned long)timer.tv_nsec, "nanosecond");
        printf("\n");
    }

    // Sleep
    while (nanosleep(&timer, &timer) != 0 && errno == EINTR) {
        if (signal_wait || current_signal == SIGALRM) {
            if (verbose)
                printf("Got signal %s!\n", strsignal(current_signal));
            return 0;
        }

        // Show remaining seconds and nanoseconds
#if defined(SIGINFO) || defined(SIGPWR)
        if(current_signal == SIGINFO || current_signal == SIGPWR)
            printf("Remaining seconds: %ld\n"
                   "Remaining nanoseconds: %ld\n",
                   (long)timer.tv_sec, timer.tv_nsec);
#endif
    }

    // Sleep for multiple centuries (workaround for 32-bit int)
    while (centuries != 0) {
        sleep((unsigned int)year * 100);
        --centuries;
    }

    // Notify the user on completion
    if (verbose)
        printf("Time's up!\n");
    return 0;
}
